compiler/p/codegen/ControlFlowEvaluator.cpp

```diff
@@ -82,71 +82,423 @@ static void lookupScheme2(TR::Node *node, bool unbalanced, bool fromTableEval, T
+   TR::InstOpCode::Mnemonic cmpiOp;
+
+   bool is64Bit;
+
+   switch (compareInfo.type)
+      {
+      case TR::Int8:
+      case TR::Int16:
+      case TR::Int32:
+         is64Bit = false;
+         break;
+      case TR::Int64:
+         is64Bit = true;
+         break;
+      case TR::Address:
+         is64Bit = cg->comp()->target().is64Bit();
+         break;
+      default:
+         TR_ASSERT_FATAL_WITH_NODE(node, false, "Cannot call evaluateIntCompareToConditionRegister with data type %s", TR::DataType::getName(compareInfo.type));
+      }
+
+   if (is64Bit && !cg->comp()->target().is64Bit())
+      return evaluateDualIntCompareToConditionRegister(condReg, node, firstChild, secondChild, compareInfo, cg);
+
+   if (is64Bit)
+      {
+      if (compareInfo.isUnsignedOrUnordered)
+         {
+         cmpOp = TR::InstOpCode::cmpl8;
+         cmpiOp = TR::InstOpCode::cmpli8;
+         }
+      else
+         {
+         cmpOp = TR::InstOpCode::cmp8;
+         cmpiOp = TR::InstOpCode::cmpi8;
+         }
+      }
    else
-      return TR::TreeEvaluator::ifiucmpleEvaluator(node, cg);
+      {
+      if (compareInfo.isUnsignedOrUnordered)
+         {
+         cmpOp = TR::InstOpCode::cmpl4;
+         cmpiOp = TR::InstOpCode::cmpli4;
+         }
+      else
+         {
+         cmpOp = TR::InstOpCode::cmp4;
+         cmpiOp = TR::InstOpCode::cmpi4;
+         }
+      }
+
+   TR::Register *firstReg = evaluateAndExtend(firstChild, compareInfo.isUnsignedOrUnordered, false, cg);
+   bool canUseCmpi = secondChild->getOpCode().isLoadConst() &&
+      (compareInfo.isUnsignedOrUnordered
+         ? is16BitUnsigedImmediate(secondChild->get64bitIntegralValueAsUnsigned())
+         : is16BitSignedImmediate(secondChild->get64bitIntegralValue()));
+
+   if (canUseCmpi)
+      {
+      generateTrg1Src1ImmInstruction(cg, cmpiOp, node, condReg, firstReg, secondChild->get64bitIntegralValue());
+      }
+   else
+      {
+      TR::Register *secondReg = evaluateAndExtend(secondChild, compareInfo.isUnsignedOrUnordered, false, cg);
+      generateTrg1Src2Instruction(cg, cmpOp, node, condReg, firstReg, secondReg);
+      stopUsingExtendedRegister(secondReg, secondChild, cg);
+      }
+
+   stopUsingExtendedRegister(firstReg, firstChild, cg);
+
+   return compareInfo.cond;
    }
 
-TR::Register *OMR::Power::TreeEvaluator::acmpltEvaluator(TR::Node *node, TR::CodeGenerator *cg)
+CompareCondition evaluateFloatCompareToConditionRegister(
+   TR::Register *condReg,
+   TR::Node *node,
+   TR::Node *firstChild,
+   TR::Node *secondChild,
+   const CompareInfo& compareInfo,
+   TR::CodeGenerator *cg
+)
+   {
+   CRCompareCondition crCond = compareConditionInCR(compareInfo.cond);
+   TR::Register *firstReg = cg->evaluate(firstChild);
+   TR::Register *secondReg = cg->evaluate(secondChild);
+
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::fcmpu, node, condReg, firstReg, secondReg);
+   if (crCond.isReversed != compareInfo.isUnsignedOrUnordered)
```

Comment:
Sorry, I afraid I don't understand this logic for using crxor. Could you please clarify?

Reply:
Sure. In general, this code is meant to solve the problem of the Power ISA thinking of `x >= y` as being equivalent to `!(x < y)` in its mnemonics, which is fine for integers but problematic for floating-point where it's possible for the two operands to be unordered. In effect, we sometimes need to add an extra `|| is_unordered(x, y)` to our check to correct for this, i.e. `x >= y -> !(x < y || is_unordered(x, y))`. The same problem applies in reverse when we specifically want unordered operands to return true.

To correct for this, we just need to OR/XOR the two bits of the CR together before returning. Since these two bits can never be set simultaneously, either operation is correct and which one is a better representation of what's happening depends on how you think about what exactly this code is doing. I'm open to switching this to use `cror` if you think it'd make it more clear.

Reply:
Right, I think it works.

Reply:
But I think a comment might be helpful.

Reply:
I've added a comment to clarify what's going on here.